### PR TITLE
Add actual computation of fee-per-cu

### LINKED ISSUE #24615: Add actual computation of transaction weight/fee-per-cu
#### Problem
This depends on #24613 landing. 

In #24613 we restructured `BankingStage` to support a heap that organizes packets by weight. However, the weight of all the packets is equal in this implementation (other than by sender stake). We need to actually compute the fee-per-cu using additional fees and base fees.

### PR DESCRIPTION
**Pull request by taozhu-chicago**
#### Problem

to compute fee-per-cu from compute_budget instruction.

#### Summary of Changes

- Use bank to sanitize versioned_message
- Compute requested_additional_fee/requested_cu from ComputeBudget instruction if tx_wide_compute_cap feature is enabled.

Fixes #24615

### CODE REVIEW

File: core/src/unprocessed_packet_batches.rs [hunk 1]
```diff
@@ -368,9 +374,40 @@ pub fn packet_message(packet: &Packet) -> Result<&[u8], DeserializedPacketError>
...
-fn compute_fee_per_cu(_message: &VersionedMessage, _bank: &Bank) -> u64 {
-    1
+/// Computes `(addition_fee / requested_cu)` if feature tx_wide_compute_cap is enabled,
+/// otherwise return default 0u64
+fn compute_fee_per_cu(message: &VersionedMessage, bank: &Bank) -> u64 {
+    if bank.feature_set.is_active(&tx_wide_compute_cap::id()) {
+        if let Some(sanitized_message) = sanitize_message(message, bank) {
+            let mut compute_budget = ComputeBudget::default();
+            match compute_budget.process_message(&sanitized_message, false, true) {
+                Ok(requested_additional_fee) => requested_additional_fee / compute_budget.max_units,
```
Comment (taozhu-chicago) [inline core/src/unprocessed_packet_batches.rs:384]: Running tests with `request_additional_fee = 3`, but the prioritization didn't happen, only to realize fee-per-cu would be `0` if requested additional_fee is less than requested units, which is 200K by default. So fee-per-cu needs to be float (don't like float), or scale up (risk overflow). debating

File: core/src/unprocessed_packet_batches.rs [hunk 2]
```diff
@@ -368,9 +374,40 @@ pub fn packet_message(packet: &Packet) -> Result<&[u8], DeserializedPacketError>
...
 }
 
-/// Computes `(addition_fee + base_fee / requested_cu)` for `deserialized_packet`
-fn compute_fee_per_cu(_message: &VersionedMessage, _bank: &Bank) -> u64 {
-    1
+/// Computes `(addition_fee / requested_cu)` if feature tx_wide_compute_cap is enabled,
+/// otherwise return default 0u64
+fn compute_fee_per_cu(message: &VersionedMessage, bank: &Bank) -> u64 {
+    if bank.feature_set.is_active(&tx_wide_compute_cap::id()) {
+        if let Some(sanitized_message) = sanitize_message(message, bank) {
```
Comment (taozhu-chicago) [inline core/src/unprocessed_packet_batches.rs:381]: And, still needs a `bank` as `address_loader` to sanitize message. Is there a way to do the same without needing a bank?

File: core/src/unprocessed_packet_batches.rs (hunk 1 above)
Comment (carllin) [reply core/src/unprocessed_packet_batches.rs:384]: what are the units currently for `requested_additional_fee`? If it's using lamports it should be much larger than the default compute units. 

We can also just multiply all the fees by a constant magnifying factor (the max compute units for instance) to account for any future fluctuations in SOL price

Comment (carllin) [reply core/src/unprocessed_packet_batches.rs:384]: If we do scale up, we should also add a test here if we're using constants like the max compute units that we can never overflow.

File: core/src/unprocessed_packet_batches.rs [hunk 3]
```diff
@@ -368,9 +374,40 @@ pub fn packet_message(packet: &Packet) -> Result<&[u8], DeserializedPacketError>
         .ok_or(DeserializedPacketError::SignatureOverflowed(sig_size))
 }
 
-/// Computes `(addition_fee + base_fee / requested_cu)` for `deserialized_packet`
-fn compute_fee_per_cu(_message: &VersionedMessage, _bank: &Bank) -> u64 {
-    1
+/// Computes `(addition_fee / requested_cu)` if feature tx_wide_compute_cap is enabled,
+/// otherwise return default 0u64
+fn compute_fee_per_cu(message: &VersionedMessage, bank: &Bank) -> u64 {
+    if bank.feature_set.is_active(&tx_wide_compute_cap::id()) {
```
Comment (carllin) [inline core/src/unprocessed_packet_batches.rs:380]: is this feature gate necessary?

File: core/src/unprocessed_packet_batches.rs [hunk 4]
```diff
@@ -368,9 +374,40 @@ pub fn packet_message(packet: &Packet) -> Result<&[u8], DeserializedPacketError>
...
+        VersionedMessage::V0(message) => {
+            let loaded_addresses = address_loader
+                .load_addresses(&message.address_table_lookups)
+                .ok()?;
+            Some(SanitizedMessage::V0(v0::LoadedMessage::new(
+                message.clone(),
+                loaded_addresses,
+            )))
+        }
+    }
```
Comment (carllin) [inline core/src/unprocessed_packet_batches.rs:410]: If we're going to sanitize every transaction here, can we avoid sanitizing when we process in `process_packets_transactions` https://github.com/carllin/solana/blob/1dc2780289c21bd1efce43ccb49021a49a086c44/core/src/banking_stage.rs#L1830-L1835

Comment (carllin) [reply core/src/unprocessed_packet_batches.rs:410]: nvm, it seems like these `SanitizedMessages` are specific to the context of a specific bank/slot: https://github.com/jeffwashington/solana/blob/a44f3c4dbeb5d8740d2029df5e26f71852be0a9c/runtime/src/accounts.rs#L574-L578, i.e. loading from a different slot could yield a different account

File: core/src/unprocessed_packet_batches.rs (hunk 2 above)
Comment (carllin) [reply core/src/unprocessed_packet_batches.rs:381]: From my investigation into the address loader here: https://github.com/solana-labs/solana/pull/25007/files#r866276411, it doesn't seem like it makes sense to sanitize the message until we get to actually processing the transaction within the leader bank, because sanitize could return different results based on which bank is being used to lookup the account. 

Double checking this is true with @jstarry

Comment (carllin) [reply core/src/unprocessed_packet_batches.rs:381]: I think we should start charging for all of these errors, AccountNotFound, InsufficientFundsForFee, InvalidAccountForFee instead of ignoring them

Then we don't need to sanitize.

File: core/src/unprocessed_packet_batches.rs (hunk 1 above)
Comment (taozhu-chicago) [reply core/src/unprocessed_packet_batches.rs:384]: Since it is fee-per-cu, I am assuming it is in lamports, as in "how many lamports per cu". Can use inputs from @jackcmay. `Additional_fee` is defined as u32 has MAX `4294967295` for a transaction, if the transaction use default 200,000 units, when user can bid up to 21,474 lamports per unit, that may not be enough?

File: core/src/unprocessed_packet_batches.rs (hunk 3 above)
Comment (taozhu-chicago) [reply core/src/unprocessed_packet_batches.rs:380]: if the feature is not enabled, then it doesn't need to sanitize the message at all.

File: core/src/unprocessed_packet_batches.rs (hunk 2 above)
Comment (taozhu-chicago) [reply core/src/unprocessed_packet_batches.rs:381]: on the same page. For the context, the reason to sanitize is so compute_budget can process the sanitized message.

File: core/src/unprocessed_packet_batches.rs (hunk 1 above)
Comment (jackcmay) [reply core/src/unprocessed_packet_batches.rs:384]: Additional fee is the total additional lamports the payer is willing to pay, it is separate from CU requested. 

There Is a mechanism in the bank to calculate a base fee based on the number of compute units requested and right now it's set to match the current per-sig fee.  That can be changed anytime with a feature gate to any quantized fee per n units.  The total fee for the tx would then be the fee per n units + additional fee

Comment (aeyakovenko) [reply core/src/unprocessed_packet_batches.rs:384]: we can prioritize based on additional_fee only.  I think it might be worth it to set the base fee to some fixed amount thats the same for all txs.

Comment (taozhu-chicago) [reply core/src/unprocessed_packet_batches.rs:384]: @jackcmay since we are prioritize only by additional_fee, what do you think to re-define `additional_fee` to `additional_lamports_per_cu`? Would make it cleaner?

Comment (jackcmay) [reply core/src/unprocessed_packet_batches.rs:384]: What if we change additional_fee to u64 and call it additional_fee_lamports?  Would require a change to the compute budget instruction and would delay the enabling of tx-wide cap on mnb...

I'm also fine with additional_lamports_per_cu, if so, should rename it in the instruction

Comment (taozhu-chicago) [reply core/src/unprocessed_packet_batches.rs:384]: Yea, both will works. Am debating with @aeyakovenko about it. 

With first one, we will have to deal with precision issues come with floats (eg, 2 TXs both has 1.4M CUs, one has additional_fee of `1`, other has `2`, comparing their fee-per-cu wouldn't be accurate, but might be OK). 

With second option, if we want the smallest fee bump to be less than 1 lamports-per-cu, say 0.001 lamport-per-cu, we then need some multiplier / factor. Could add confusion to user. 

Let's take a vote 😃

Comment (jackcmay) [reply core/src/unprocessed_packet_batches.rs:384]: a difference of 1 or 2 lamports is pretty small difference

Comment (jackcmay) [reply core/src/unprocessed_packet_batches.rs:384]: How about this?
https://github.com/solana-labs/solana/pull/25027

Comment (taozhu-chicago) [reply core/src/unprocessed_packet_batches.rs:384]: > a difference of 1 or 2 lamports is pretty small difference

just an example 😜

Comment (taozhu-chicago) [reply core/src/unprocessed_packet_batches.rs:384]: > How about this? #25027

Looks good, can take `u64` approach

Comment (carllin) [reply core/src/unprocessed_packet_batches.rs:384]: f64's for fee-per-cu seem fine?

f64 has 15 decimal precision, so on a 1.4m compute unit transaction you would only need 1.4 * 10^-9 lamports increase on your total fee to move the fee-per-cu. This is well below the 1 lamport incremental units specified in the additional fee

File: core/src/unprocessed_packet_batches.rs (hunk 3 above)
Comment (carllin) [reply core/src/unprocessed_packet_batches.rs:380]: Seems like we can parse the additional fee without sanitizing the messagefirst

All we need is the `message` from the `VersionedTransaction`: https://github.com/solana-labs/solana/blob/master/sdk/src/transaction/sanitized.rs#L97 and then parse the instructions here: https://github.com/solana-labs/solana/blob/master/program-runtime/src/compute_budget.rs#L112

Comment (taozhu-chicago) [reply core/src/unprocessed_packet_batches.rs:380]: but should we parse without sanitizing first?

File: core/src/unprocessed_packet_batches.rs (hunk 1 above)
Comment (taozhu-chicago) [reply core/src/unprocessed_packet_batches.rs]: commit b2c17755d887e0c5bcd0b375c25260997c6e3bb9 to change fee-per-cu to f64.

File: core/src/unprocessed_packet_batches.rs (hunk 3 above)
Comment (taozhu-chicago) [reply core/src/unprocessed_packet_batches.rs:385]: you mean to simplify `fn sanitize_message(versioned_message)`  by doing: 
1. versioned_message.sanitize().ok()?; // still good to sanitize message before using it
2. straight process `message` from either legacy or V0 for compute_budget instruction? With some copied code?

Comment (carllin) [reply core/src/unprocessed_packet_batches.rs:385]: ok so after a helpful chat with @jstarry and @t-nelson you're right, you can't parse it out without the sanitize. Sorry for the confusion!

But this does mean we incur an extra account lookup for every transaction being added to the queue, potentially perf degrading

File: core/src/unprocessed_packet_batches.rs (hunk 2 above)
Comment (t-nelson) [reply core/src/unprocessed_packet_batches.rs:386]: sanitize for `v0::Message`s can incur multiple account lookups to resolve address maps. putting this before tx scheduling may introduce a DOS vector.

further, as Carl mentioned, `v0::Message`s expect address maps to be resolved and the transaction executed on the the same `Bank`, we no longer have that guarantee with this change.

Justin is going to work on requiring that top-level instruction program ids be static addresses, declared directly in the transaction rather than via address maps.  This will allow us to defer address map resolution to transaction accounts load and resolve ComputeBudget instructions via static analysis.

File: core/src/unprocessed_packet_batches.rs (hunk 1 above)
Comment (t-nelson) [reply core/src/unprocessed_packet_batches.rs]: am i the only one concerned that prioritizing based purely on `additional_fee` is going to create perverse incentives?  they encode nothing except, "user desire to execute a transaction," which has nothing to do with resource contention within the system.  that is, users have to pay the price for _every_ resource under contention, regardless of whether their transaction makes use of it.  otherwise independent resource contentions become additive, everyone loses.

this seems to me the opposite of what the cluster economics should be trying to do, which is to minimize negative externalities of resource contention. containing fee rise to individual resource puts pressure on dapps to optimize their code and/or institute internal economic games. a Good Thing :tm:.  i recognize that we aren't in a position to implement anything like this today, but i don't want to see its possibility painted out of the picture out of desperation.

never mind there being
* zero client-side support for `additional_fee`, basically handing the chain over to (bot) devs in the meantime
* no tooling to inform effective `additional_fee` values
* a fairly non-trivial to resolve [footgun](https://github.com/solana-labs/solana/issues/24612) stemming from the fact that `addition_fee` is under the transaction hash, preventing re-submission of the otherwise same transaction with a higher fee without jumping through hoops or risking both submissions being executed.

File: core/src/unprocessed_packet_batches.rs (hunk 2 above)
Comment (jstarry) [reply core/src/unprocessed_packet_batches.rs:386]: My PRs are here:
https://github.com/solana-labs/solana/pull/25035
https://github.com/solana-labs/solana/pull/25041

File: core/src/unprocessed_packet_batches.rs [hunk 5]
```diff
@@ -155,13 +159,14 @@ impl PartialOrd for ImmutableDeserializedPacket {
 
 impl Ord for ImmutableDeserializedPacket {
     fn cmp(&self, other: &Self) -> Ordering {
-        match self.fee_per_cu().cmp(&other.fee_per_cu()) {
+        match self.fee_per_cu().partial_cmp(&other.fee_per_cu()).unwrap_or(Ordering::Equal) {
```
Comment (carllin) [inline core/src/unprocessed_packet_batches.rs:162]: seems like these comparisons should never return `None`: https://doc.rust-lang.org/src/core/cmp.rs.html#1396-1398, so we can unwrap on all these

File: core/src/unprocessed_packet_batches.rs (hunk 2 above)
Comment (jstarry) [reply core/src/unprocessed_packet_batches.rs:386]: @taozhu-chicago you can rebase off of #25041 to use `SanitizedVersionedMessage` here

File: core/src/unprocessed_packet_batches.rs [hunk 6]
```diff
@@ -368,9 +379,48 @@ pub fn packet_message(packet: &Packet) -> Result<&[u8], DeserializedPacketError>
...
+/// Computes `(addition_fee / requested_cu)` if feature tx_wide_compute_cap is enabled,
+/// otherwise return default 0u64
+fn compute_fee_per_cu(message: &VersionedMessage, bank: &Bank) -> f64 {
+    if bank.feature_set.is_active(&tx_wide_compute_cap::id()) {
+        if let Some(sanitized_message) = sanitize_message(message, bank) {
+            let mut compute_budget = ComputeBudget::default();
+            match compute_budget.process_message(&sanitized_message, false, true) {
+                Ok(requested_additional_fee) => {
+                    if let Some(requested_units) = f64::from_u64(compute_budget.max_units) {
+                        let requested_additional_fee = f64::from_u64(requested_additional_fee).unwrap_or(0f64);
```
Comment (carllin) [inline core/src/unprocessed_packet_batches.rs:391]: do we have any sanitization to make sure the `requested_additional_fee` isn't too high? This conversion could fail and then we crowd the buffer with a bunch of `requested_additional_fee == 0`

File: core/src/unprocessed_packet_batches.rs (hunk 5 above)
Comment (taozhu-chicago) [reply core/src/unprocessed_packet_batches.rs]: ah, done

File: core/src/unprocessed_packet_batches.rs (hunk 2 above)
Comment (taozhu-chicago) [reply core/src/unprocessed_packet_batches.rs:385]: awesome, cherry-picked

Comment (taozhu-chicago) [reply core/src/unprocessed_packet_batches.rs]: commit 5e43496762a9421799626318aaf805a4fb888bf2 uses `SanitizedVersionedMessage` to process compute_budget, without needing a `bank`! Awesome ... thanks @jstarry

File: core/src/unprocessed_packet_batches.rs (hunk 6 above)
Comment (taozhu-chicago) [reply core/src/unprocessed_packet_batches.rs]: commit 5e43496762a9421799626318aaf805a4fb888bf2 simplified `compute_fee_per_cu` that now returns an Option<f64>,

File: core/src/unprocessed_packet_batches.rs (hunk 1 above)
Comment (taozhu-chicago) [reply core/src/unprocessed_packet_batches.rs]: @aeyakovenko @carllin what's your take on @t-nelson's concern?

Using `additional_fee / requested_cu` to prioritize transactions gives user a tool to increase chance of landing their TXs regardless if there's resource contention, but especially useful during contention. Believe it's a real possibility that bots takes advantage of this before real user, but the fee should discourage bots going too far, would it?

Comment (nikhayes) [reply core/src/unprocessed_packet_batches.rs]: > Additional fee is the total additional lamports the payer is willing to pay, it is separate from CU requested.
> 
> There Is a mechanism in the bank to calculate a base fee based on the number of compute units requested and right now it's set to match the current per-sig fee. That can be changed anytime with a feature gate to any quantized fee per n units. The total fee for the tx would then be the fee per n units + additional fee

Hmm might it be easier if it's just price of n units * whole number fee multiplier? I feel like having an additional fee that can be incremented to a tiny degree could result in a game between bots trying to minutely increase fee priority that results in an increase in spam. It might also be easier for UX because users could choose fee multiplier defaults in the wallet.

Comment (taozhu-chicago) [reply core/src/unprocessed_packet_batches.rs]: the `whole number multiplier` sounds similar to `additional_lamports_per_cu` suggested [early](https://github.com/solana-labs/solana/pull/25007#discussion_r866363339). I like this idea for its simplicity.

**Comment by taozhu-chicago** [comment]
closing this in favor of #25122 which rebased and picked #25067
